File: reasoning/python/graph.py
# See https://en.wikipedia.org/wiki/Graph_theory

if __name__ == "__main__":
    oneway = [
        ('paris', 'orleans'),
        ('paris', 'chartres'),
        ('paris', 'amiens'),
        ('orleans', 'blois'),
        ('orleans', 'bourges'),
        ('blois', 'tours'),
        ('chartres', 'lemans'),
        ('lemans', 'angers'),
        ('lemans', 'tours'),
        ('angers', 'nantes')
    ]

    path = []

    path += oneway

    while True:
        ol = len(path)
        for (A, B) in oneway:
            for (C, D) in path:
                if B == C and (A, D) not in path:
                    path += [(A, D)]
        if len(path) == ol:
            break

    cases = [
        "path"
    ]

    for c in cases:
        print('[] :python-answer """%s = %s""".' % (c, eval(c)))

File: looking/julia/graph.jl
# See https://en.wikipedia.org/wiki/Graph_(discrete_mathematics)

using Julog

clauses = @julog [
    oneway(paris, orleans) <<= true,
    oneway(paris, chartres) <<= true,
    oneway(paris, amiens) <<= true,
    oneway(orleans, blois) <<= true,
    oneway(orleans, bourges) <<= true,
    oneway(blois, tours) <<= true,
    oneway(chartres, lemans) <<= true,
    oneway(lemans, angers) <<= true,
    oneway(lemans, tours) <<= true,
    oneway(angers, nantes) <<= true,

    path(X, Y) <<= oneway(X, Y),
    path(X, Y) <<= oneway(X, Z) & path(Z, Y)
]

cases = [
    :(resolve(@julog(path(X, angers)), clauses)),
    :(resolve(@julog(path(angers, X)), clauses)),
    :(resolve(@julog(path(paris, nantes)), clauses))
]

for case in cases
    println("[ :julia-result \"", case, " = ", eval(case), "\"].")
end

